core/src/main/scala/cats/data/Coproduct.scala

```diff
@@ -0,0 +1,217 @@
+package cats
+package data
+
+import cats.functor.Contravariant
+
+/** `F` on the left and `G` on the right of [[Xor]].
+  *
+  * @param run The underlying [[Xor]]. */
+final case class Coproduct[F[_], G[_], A](run: F[A] Xor G[A]) {
+
+  import Coproduct._
+
+  def map[B](f: A => B)(implicit F: Functor[F], G: Functor[G]): Coproduct[F, G, B] =
+    Coproduct(run.bimap(F.lift(f), G.lift(f)))
+
+  def coflatMap[B](f: Coproduct[F, G, A] => B)(implicit F: CoflatMap[F], G: CoflatMap[G]): Coproduct[F, G, B] =
+    Coproduct(
+      run.bimap(a => F.coflatMap(a)(x => f(leftc(x))), a => G.coflatMap(a)(x => f(rightc(x))))
+    )
+
+  def duplicate(implicit F: CoflatMap[F], G: CoflatMap[G]): Coproduct[F, G, Coproduct[F, G, A]] =
+    Coproduct(run.bimap(
+      x => F.coflatMap(x)(a => leftc(a))
+      , x => G.coflatMap(x)(a => rightc(a)))
+    )
+
+  def extract(implicit F: Comonad[F], G: Comonad[G]): A =
+    run.fold(F.extract, G.extract)
+
+  def contramap[B](f: B => A)(implicit F: Contravariant[F], G: Contravariant[G]): Coproduct[F, G, B] =
+    Coproduct(run.bimap(F.contramap(_)(f), G.contramap(_)(f)))
+
+  def foldRight[B](z: Eval[B])(f: (A, Eval[B]) => Eval[B])(implicit F: Foldable[F], G: Foldable[G]): Eval[B] =
+    run.fold(a => F.foldRight(a, z)(f), a => G.foldRight(a, z)(f))
+
+  def foldLeft[B](z: B)(f: (B, A) => B)(implicit F: Foldable[F], G: Foldable[G]): B =
+    run.fold(a => F.foldLeft(a, z)(f), a => G.foldLeft(a, z)(f))
+
+  def foldMap[B](f: A => B)(implicit F: Foldable[F], G: Foldable[G], M: Monoid[B]): B =
+    run.fold(F.foldMap(_)(f), G.foldMap(_)(f))
+
+  def traverse[X[_], B](g: A => X[B])(implicit F: Traverse[F], G: Traverse[G], A: Applicative[X]): X[Coproduct[F, G, B]] =
+    run.fold(
+      x => A.map(F.traverse(x)(g))(leftc(_))
+      , x => A.map(G.traverse(x)(g))(rightc(_))
+    )
+
+  def isLeft: Boolean =
+    run.isLeft
+
+  def isRight: Boolean =
+    run.isRight
+
+  def swap: Coproduct[G, F, A] =
+    Coproduct(run.swap)
+
+  def toValidated: Validated[F[A], G[A]] =
+    run.toValidated
+
+}
+
+object Coproduct extends CoproductInstances {
+
+  def leftc[F[_], G[_], A](x: F[A]): Coproduct[F, G, A] =
+    Coproduct(Xor.left(x))
+
+  def rightc[F[_], G[_], A](x: G[A]): Coproduct[F, G, A] =
+    Coproduct(Xor.right(x))
+
+  final class CoproductLeft[G[_]] private[Coproduct] {
+    def apply[F[_], A](fa: F[A]): Coproduct[F, G, A] = Coproduct(Xor.left(fa))
+  }
+
+  final class CoproductRight[F[_]] private[Coproduct] {
+    def apply[G[_], A](ga: G[A]): Coproduct[F, G, A] = Coproduct(Xor.right(ga))
+  }
+
+  def left[G[_]]: CoproductLeft[G] = new CoproductLeft[G]
+
+  def right[F[_]]: CoproductRight[F] = new CoproductRight[F]
+
+}
+
+sealed abstract class CoproductInstances3 {
```

Comment:
No worries, I want to get it right so I understand what the expectations are for future contributions.:) Added those access modifiers also on `InjectInstances`